### ISSUE TITLE
SSIM loss问题

### ISSUE THREAD

**Issue by gegrace-1210** [G-pz/HPN-CR#4]
您好，请问训练过程中，使用的pytorch_msssim.MS_SSIM(data_range=1, channel=12)这个loss在训练一段时间后就会出现NAN，请问如何解决？已经排查了输入的lable和pre图像值域均在0-1且不是nan，前面loss都在正常下降，但会训练到第92轮次 SSIM loss为nan，如果降低lr，轮次也将增加，在三百多轮SSIM loss变为nan.

**Comment by G-pz** [comment]
It is recommended to refer to the data processing methods and hyperparameter settings described in the paper. Additionally, you may also consider common approaches for handling gradient explosion.

**Comment by try-agaaain** [comment]
我遇到了类似的问题，GPT提供了一个有效的建议：

在 loss.py 文件中，`_sam` 函数计算光谱角映射（SAM），其中 `torch.acos()` 对浮点数精度问题很敏感。如果其输入由于计算误差略微超出 `[-1, 1]` 的范围，就会返回 `NaN`。此外，分母也可能为零。

**修改 loss.py 文件：**
在计算 `acos` 之前，使用 `torch.clamp` 将输入限制在有效范围内，并为分母添加一个极小值 `eps` 以防止除以零。

````python
// ...existing code...
def _sam(x1, x2):

    B,N,_,_ = x1.shape
    x1_ = x1.reshape(B * N, -1)
    x2_ = x2.reshape(B * N, -1)

    eps = 1e-6  # Add epsilon for numerical stability
    norm1 = torch.sqrt(torch.sum(x1_ ** 2, axis=1))
    norm2 = torch.sqrt(torch.sum(x2_ ** 2, axis=1))
    
    A = torch.sum(x1_ * x2_, axis=1) / (norm1 * norm2 + eps)
    A = torch.clamp(A, -1, 1)  # Clamp the input to acos to prevent NaN
    
    SAM = torch.acos(A) * 180 / math.pi

    return torch.mean(SAM)

class SAMLoss(nn.Module):
// ...existing code...
````
